### ISSUE TITLE
RFC:  Docker image versioning + OpenVPN releases

### ISSUE THREAD

**Issue by kylemanna** [kylemanna/docker-openvpn#267]
Hey All,

With the release of OpenVPN 2.4 alot of people have requested it.  For a while I had hoped that Alpine would update their stable 3.5 branch to it and it would get rolled in to the image automatically, but this has not happened just yet.

I'm interested in hearing what people thing about ways to accomplish the following:
1. Provide a stable release under the default Docker `latest` tag.  Something people can blindly use with little risk.
2. Provide Docker tags that track an upstream release, both past releases and new releases.

To accomplish goal (1) I assume the master branch will just proceed as it is with no surprises.  At some point Alpine 3.6 will come out and it will update at that point.

For goal (2) past releases, I propose stable branches that track OpenVPN release.  For example branch `openvpn-v2.3` with corresponding docker tag `2.3`.  These branches likely will be maintained only when people notice or find issues.

To accomplish (2) for future releases, there may be an `edge` branch and corresponding Docker tag.  This will track the Alpine Edge release and have the latest Alpine openvpn release.  Anyone wanting the latest features can follow this.

To recap:
1. The git `master` branch and Docker `latest` tag stay as they are today.
2. I'll create an `edge` branch and Docker `edge` tag that will use `alpine:edge` as the base Docker image.
3. I'll create a `openvpn-v2.3` branch with the current `master` HEAD and a corresponding Docker `2.3` tag for those running deployments to adopt for stability.

Thoughts?

**Comment by kylemanna** [comment]
Related: #1, #163, #262

**Comment by pieterlange** [comment]
Sounds good to me.

**Comment by chepurko** [comment]
Yes, good.

**Comment by breznak** [comment]
Also related to #264 

Definitely :+1: for the release branches/tags and for providing `edge` with newer releases. 

- I'd even consider `upstream`/`git` that would provide the development -git version of OpenVPN
- about the ways to deal with it: 
  - is `AlpineLinux` a hard requirement? (I don't know how many benefits does it have for this project, compared to other distros). Could we use other distro that has OpenVPN better (faster) maintained? 
  - Contribute to  Alpine and provide them the package? But I guess the `edge` solves it, may be applicable for the `git` versions. 
  - Why not build the OVPN package ourselves? And then include only the binary (and deps) in the image?

**Comment by kylemanna** [comment]
> I'd even consider upstream/git that would provide the development -git version of OpenVPN

I don't think anyone wants this.  Anyone who might want the bleeding edge releases should be more then capable of submitting a pull request themselves.  Nobody has brought this up, and nobody has submitted a pull request.  I think that the few days that Alpine Linux lags on OpenVPN releases is more then sufficient.

Only a handful of users have even requested OpenVPN 2.4.x and it has been out for some time.

**Comment by buchdag** [comment]
@kylemanna sounds good to me too.

**Comment by DittmarSteiner** [comment]
I’d definitely like to have OpenVPN 2.4.x for coming roll-outs, where I don’t want to upgrade later.
Thank you!

**Comment by kylemanna** [comment]
[Alpine 3.6.0 was just released](https://www.alpinelinux.org/posts/Alpine-3.6.0-released.html) and [OpenVPN 2.4.2](https://pkgs.alpinelinux.org/packages?name=openvpn&branch=&repo=&arch=x86_64&maintainer=) is included.

**Comment by pieterlange** [comment]
alpine:3.6 docker image is available now too :+1:

**Comment by pieterlange** [comment]
Updated within 24 hours. Thanks for the amazing support on the project @kylemanna! The new features provided in 2.4 will surely prove to be useful!

---
 * There's a typo in the original issue body: 4.2 instead of 2.4
 * I think i fixed the `addArg` function [here](https://github.com/pieterlange/kube-openvpn/commit/17c2e043f9edfc86b981453fd6bda40cb4945268) in order to support using arguments with multiple options. Only ran into this after upgrading to alpine:3.6, so i guess bash changed a bit.

**Comment by kylemanna** [comment]
Updated as described above, but moved `master` branch to `alpine:latest`  which will bring master branch to openvpn 2.4.

**Comment by kylemanna** [comment]
> I think i fixed the addArg function here in order to support using arguments with multiple options. Only ran into this after upgrading to alpine:3.6, so i guess bash changed a bit.

@pieterlange does this trip up some of the tests?

**Comment by pieterlange** [comment]
Probably.

Might have something to do with the `exec` invocation in my `entrypoint.sh` as well.

**Comment by kylemanna** [comment]
The Travis-CI tests passed.  If there's a problem, please open an issue or PR and we should fix the tests so that they detect it.

Closing this issue for now.  Thanks for the feedback and comments all!
